events/sources/kubernetes/kubernetes_source.go

```diff
@@ -0,0 +1,151 @@
+	localEventsBuffer chan *kubeapi.Event
+
+	stopChannel  chan struct{}
+	errorChannel chan error
+
+	eventClient kubeclient.EventInterface
+}
+
+func (this *KubernetesEventSource) GetNewEvents() *core.EventBatch {
+	result := core.EventBatch{
+		Timestamp: time.Now(),
+		Events:    []*kubeapi.Event{},
+	}
+	// Get all data from the buffer.
+	for {
+		select {
+		case event := <-this.localEventsBuffer:
+			result.Events = append(result.Events, event)
+		default:
+			break
+		}
+	}
+	return &result
+}
+
+func (this *KubernetesEventSource) watch() {
+	defer close(this.errorChannel)
+
+	// Outer loop, for reconnections.
+	for {
+		events, err := this.eventClient.List(kubelabels.Everything(), kubefields.Everything())
+		if err != nil {
+			glog.Fatalf("Failed to load events: %v", err)
+			this.errorChannel <- fmt.Errorf("Failed to load events")
+			return
+		}
+		// Do not write old events.
+
+		resourceVersion := events.ResourceVersion
+
+		watcher, err := this.eventClient.Watch(
+			kubelabels.Everything(),
+			kubefields.Everything(),
+			kubeapi.ListOptions{
+				LabelSelector:   kubelabels.Everything(),
+				FieldSelector:   kubefields.Everything(),
+				Watch:           true,
+				ResourceVersion: resourceVersion})
+		if err != nil {
+			glog.Fatalf("Failed to start watch for new events: %v", err)
+			this.errorChannel <- fmt.Errorf("Failed to start watch")
+			return
+		}
+
+		watchChannel := watcher.ResultChan()
+		// Inner loop, for update processing.
+		for {
+			select {
+			case watchUpdate, ok := <-watchChannel:
+
+				if !ok {
+					glog.Errorf("Event watch channel closed")
+					break
+				}
+
+				if watchUpdate.Type == kubewatch.Error {
+					if status, ok := watchUpdate.Object.(*kubeapiunv.Status); ok {
+						glog.Errorf("Error during watch: %#v", status)
+						break
+					}
+					glog.Errorf("Received unexpected error: %#v", watchUpdate.Object)
+					break
+				}
+
+				if event, ok := watchUpdate.Object.(*kubeapi.Event); ok {
+
+					switch watchUpdate.Type {
+					case kubewatch.Added, kubewatch.Modified:
+						select {
+						case this.localEventsBuffer <- event:
+							// Ok, buffer not full.
+						default:
+							// Buffer full, need to drop the event.
+							glog.Errorf("Event buffer full, dropping event")
+						}
+					case kubewatch.Deleted:
+						// Deleted events are silently ignored.
+					default:
+						glog.Warningf("Unknown watchUpdate.Type: %#v", watchUpdate.Type)
+					}
+				} else {
+					glog.Fatalf("Wrong object received: %v", watchUpdate)
+				}
+
+			case <-this.stopChannel:
+				glog.Infof("Event watching stopped")
+				return
+			}
+		}
+	}
+}
+
+func NewKubernetesSource(client kubeclient.Client) *KubernetesEventSource {
+	eventClient := client.Events(kubeapi.NamespaceAll)
+	result := KubernetesEventSource{
+		localEventsBuffer: make(chan *kubeapi.Event, LocalEventsBufferSize),
+		stopChannel:       make(chan struct{}),
+		errorChannel:      make(chan error, 10),
```

Comment:
Why 10? It should be either 1, as we always return after seeing an error, or more, but then we need to handle the case when it's full.